pkg/terminal/command.go

```diff
@@ -2840,29 +2840,90 @@ func printBreakpointInfo(t *Term, th *api.Thread, tracepointOnNewline bool) {
 			fmt.Fprintf(t.stdout, "\t%s: %s\n", v.Name, v.MultilineString("\t", ""))
 		}
 	}
-
 	if bpi.Stacktrace != nil {
-		tracepointnl()
-		fmt.Fprintf(t.stdout, "\tStack:\n")
-		printStack(t, t.stdout, bpi.Stacktrace, "\t\t", false)
+		if th.Breakpoint.TraceFollowCalls <= 0 {
+			tracepointnl()
+			fmt.Fprintf(t.stdout, "\tStack:\n")
+			printStack(t, t.stdout, bpi.Stacktrace, "\t\t", false)
+		}
 	}
 }
 
 func printTracepoint(t *Term, th *api.Thread, bpname string, fn *api.Function, args string, hasReturnValue bool) {
 	if t.conf.TraceShowTimestamp {
 		fmt.Fprintf(t.stdout, "%s ", time.Now().Format(time.RFC3339Nano))
 	}
+	// Simple Trace; no stack required to calculate depth of functions
+	if th.Breakpoint.TraceFollowCalls <= 0 {
+		if th.Breakpoint.Tracepoint {
+			fmt.Fprintf(t.stdout, "> goroutine(%d): %s%s(%s)\n", th.GoroutineID, bpname, fn.Name(), args)
+			printBreakpointInfo(t, th, !hasReturnValue)
+		}
+		if th.Breakpoint.TraceReturn {
+			retVals := make([]string, 0, len(th.ReturnValues))
+			for _, v := range th.ReturnValues {
+				retVals = append(retVals, v.SinglelineString())
+			}
+			fmt.Fprintf(t.stdout, ">> goroutine(%d): => (%s)\n", th.GoroutineID, strings.Join(retVals, ","))
+		}
+		if th.Breakpoint.TraceReturn || !hasReturnValue {
+			if th.BreakpointInfo != nil && th.BreakpointInfo.Stacktrace != nil {
+				fmt.Fprintf(t.stdout, "\tStack:\n")
+				printStack(t, t.stdout, th.BreakpointInfo.Stacktrace, "\t\t", false)
+			}
+		}
+		return
+	}
+
+	// Trace Follow Calls; stack is required to calculate depth of functions
+	wantindex := -1
+	rootindex := -1
+	if th.BreakpointInfo == nil || th.BreakpointInfo.Stacktrace == nil {
+		return
+	}
+
+	stack := th.BreakpointInfo.Stacktrace
```

Comment:
Can you explain what the logic behind this calculation is? Why is `wantindex` even needed? Why are you calculating curi as `len(stack) - 1 - i` when you later use it in a way that will always cancel the `len(stack)` term?

Reply:
There were many reasons for this- we needed rootindex to mark the depth of the starting func name as without that, it would report depth relative to main and not the function which was being given as a parameter, calculating it as len(stack)-1-i helps us to traverse from bottom up so that we can set rootindex properly at the first match in case the same function gets recursively invoked later (to avoid overwriting it with a higher depth etc), wantindex was required to mark depth of the function corresponding to the tracepoint.

Reply:
Right, but why is wantindex the deepest call of the current tracepoint function and rootindex the least deep call? What if the current tracepoint function is recursively calling itself? Should that not increment the depth? What if the deepest call of the current tracepoint function is deeper than the least deep call of the root function?

Reply:
To be honest I undestand what this code is trying to do so little that I can't really tell what's intended behavior and what's a bug.